File: 2021/2/main.php
<?php

require('advent.php');

class Puzzle extends Advent
{
    public function a()
    {
        $file = $this->readFileAsArray();
        $d = $h = 0;
        foreach ($file as $v) {
            $parts = explode(' ', $v);
            switch ($parts[0]) {
                case 'forward':
                    $h += (int) $parts[1];
                    break;
                case 'down':
                    $d += (int) $parts[1];
                    break;
                case 'up':
                    $d -= (int) $parts[1];
            }
        }

        return $d * $h;
    }

    public function b()
    {
        $file = $this->readFileAsArray();
        $d = $h = $a = 0;
        foreach ($file as $v) {
            $parts = explode(' ', $v);
            switch ($parts[0]) {
                case 'forward':
                    $h += (int)$parts[1];
                    $d += $a * (int)$parts[1];
                    break;
                case 'down':
                    $a += (int) $parts[1];
                    break;
                case 'up':
                    $a -= (int) $parts[1];
            }
        }

        return $d * $h;
    }
}

File: 2021/2/main.go
package main

import (
	"advent"
	"fmt"
	"strings"
	"strconv"
)

func main() {
	a()
	b()
}

func a() {
	data := tools.ReadFileAsArray("2")

	var h, d int
	for _, row := range data {
		if (row == "") {
			continue
		}

		val := strings.Fields(row)
		dist, _ := strconv.Atoi(val[1])

		switch val[0] {
		case "forward":
			h += dist
		case "down":
			d += dist
		case "up":
			d -= dist
		}
	}

	fmt.Println(h*d)
}

func b() {
	data := tools.ReadFileAsArray("2")

	var h, d, a int
	for _, row := range data {
		if (row == "") {
			continue
		}

		val := strings.Fields(row)
		dist, _ := strconv.Atoi(val[1])

		switch val[0] {
		case "forward":
			h += dist
			d += a*dist
		case "down":
			a += dist
		case "up":
			a -= dist
		}
	}

	fmt.Println(h*d)
}